updated models to incorporate APScheduler

honlnm opened the pull request:
I updated models.py to include a table for card info. I removed card information from 3 other tables, wish list, cards owned, and deck cards, and instead incorporated a foreign key/relationship for the needed data.

honlnm committed:
updated models to incorporate APScheduler

JB0925 commented on models.py:
```diff
@@ -104,15 +90,67 @@ class DeckCards(db.Model):
+    card_multiverse_id = db.Column(
+        db.Integer, db.ForeignKey("card_info.card_multiverse_id"), primary_key=True
+    )
+    cards_owned = db.Column(
```
This comment is for lines 102, 105, and 108. Assuming this is a join table, what do these columns represent? Do they correspond to an ID on another table? If so, maybe we could give them a name like `wish_list_id`, or something similar, for readability purposes.

honlnm replied:
card multiverse id is a column on the card table (the primary key). It's the card identifier that comes from the API and only exists for the cards we want to pull from the API

JB0925 commented on models.py:
```diff
@@ -104,15 +90,67 @@ class DeckCards(db.Model):
+        db.Integer, db.ForeignKey("cards_owned.id"), primary_key=True
+    )
+    wish_list = db.Column(
+        db.Integer, db.ForeignKey("card_wish_list.id"), primary_key=True
```
I am curious; I notice that this table has four primary keys - one on each column. Typically, when something like this is done, it is called a composite primary key.

To actually create a composite primary key in SQLAlchemy, an extra [step](https://medium.com/@thinesh12/unlocking-the-power-of-composite-primary-keys-in-sqlalchemy-b378fb975e9b#:~:text=Defining%20a%20Composite%20Primary%20Key,within%20your%20SQLAlchemy%20model%20class.) needs to be taken.

I am curious - do these four columns combine together to represent a unique entry? Like no one else can have this combination? I'm assuming so since they all appear to be IDs.

If so, this could be a good use of a composite primary key. It will create an index on these columns, which is an optimization for speeding up reads.

One thing to note about indexes, especially ones that group multiple columns, is that writes can be slower, because the index needs to be rebuilt each time there is a write ( `INSERT`, `UPDATE`, etc.).

However, in this app, the frequency of reads is likely to be much greater than that of writes, so a composite primary key here probably makes sense. It will help if:

1. Your queries use all four keys.
2. Typically, it can also be useful if, when you don't use all keys in a query, you at least use the leading key and maybe one of the others.

All that said, if these do not somehow uniquely identify a record, then it's probably not worth it and no worries.

honlnm replied:
I'm not 100% sure a composite primary key is necessary for this build. The reason I say that is due to loading efficiency. If the index needs to be rebuilt each time there is a write (aka, each time someone adds a card to inventory or their wish list, etc) it will make that process slower than we want it. Also - usually only one to two keys will only be used at one time.

honlnm committed:
updated models.py

honlnm committed:
flask_apscheduler

JB0925 committed:
Changes include:

1. Add new dependency, 'aiohttp', to allow for async fetching of card data.
2. Fetch card data asynchronously.
3. Run job asynchronously so as not to block server event loop.

JB0925 committed:
Changes include:

1. Decrease the number of urls per job to have more jobs running at once.

honlnm committed:
Merge pull request #9 from honlnm/async

Async branch to fork branch, Improve process efficiency of pulling card data from api

honlnm committed:
partially updated apiClient, app, requirements
